### ISSUE TITLE
Suggestions: basic/advanced/experimental patterns

### ISSUE THREAD

**Issue by SachaG** [graphql-rules/graphql-rules#25]
To make it easier to reach consensus on specific patterns, I think it could be useful to tag them according to their "status". What I have in mind would be:

- **Basic**: patterns or best practices that nearly everybody agrees on and that should probably be implemented by most GraphQL projects.
- **Advanced**: patterns or practices that people agree on, but might not be needed by every project. For example, something like cursor-based pagination. 
- **Experimental**: patterns that some people find useful, but haven't reached a consensus yet; or maybe go against the GraphQL spec somehow. 

Having this "experimental" designation would make it easier to include new patterns even if people don't unanimously approve of them. They could always "graduate" to basic or advanced later on once they become more widespread.

Thoughts?

**Comment by RIP21** [comment]
Totally agree, that would be nice. Because implementing all the amazing tricks described there is not needed for sure, but having a distinctions like that for newcomers would be amazing. 
There was also a plan to build some sort of picker mode. Where teams can pick the rules they are using, and then generate their own "cheatsheet".

PS. Thanks for state of JS ;)

**Comment by SachaG** [comment]
Picker mode would be cool! You could then generate stats to see which patterns are used the most. I expect it would require a lot more back-end complexity to track that data though…

And thanks, stay tuned for the 2019 edition soon (which I should be working on right now instead of doing this actually…)

**Comment by nodkz** [comment]
I agree with **basic** and **advanced** but don't feel that **experimental** name is good enough.

For now, we have the following set of rules. And which tags do you want to set for every rule? Maybe you will found new tags ;)

<ul>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/01-naming/README.md">1. Naming rules</a>
<ul>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/01-naming/naming-fields-args.md">1.1. Use <code>camelCase</code> for GraphQL-fields and arguments.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/01-naming/naming-types.md">1.2. Use <code>UpperCamelCase</code> for GraphQL-types.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/01-naming/naming-enum.md">1.3. Use <code>CAPITALIZED_WITH_UNDERSCORES</code> to name ENUM-types.</a></li>
</ul>
</li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/02-type/README.md">2. Type rules</a>
<ul>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/02-type/type-custom-scalars.md">2.1. Use custom scalar types if you want to declare fields or args with specific semantic value.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/02-type/type-enumerable.md">2.2. Use Enum for fields which contain a specific set of values.</a></li>
</ul>
</li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/03-fields-output/README.md">3. Field Rules (Output)</a>
<ul>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/03-fields-output/output-semantic-names.md">3.1. Use semantic names for fields and avoid leaking of implementation details in fields names.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/03-fields-output/output-non-null.md">3.2. Use <code>Non-Null</code> field if field will always have a given field value.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/03-fields-output/output-grouping.md">3.3. Group as many related fields into custom Object type as possible.</a></li>
</ul>
</li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/04-fields-input/README.md">4. Argument rules (Input)</a>
<ul>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/04-fields-input/input-grouping.md">4.1. Group coupled arguments to the new input-type.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/04-fields-input/input-custom-scalar.md">4.2.Use strict scalar types for arguments, eg. <code>DateTime</code> instead of <code>String</code>.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/04-fields-input/input-non-null.md">4.3. Mark arguments as <code>required</code>, is they are required for query execution.</a></li>
</ul>
</li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/05-list/README.md">5. Rules of lists</a>
<ul>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/05-list/list-filter.md">5.1. To filter the lists, use the <code>filter</code> argument, which contains all the available filters.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/05-list/list-sort.md">5.2. Use argument <code>sort</code> of type <code>Enum</code> or <code>[Enum!]</code> for listings sorting.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/05-list/list-limit-skip.md">5.3. Use <code>limit</code> with default value and <code>skip</code> to limit number of returned items in list.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/05-list/list-pagination.md">5.4. Use <code>page</code>, <code>perPage</code> args for pagination and return output type with <code>items</code> (array of elements) and <code>pageInfo</code> (meta-data).</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/05-list/list-cursor-pagination.md">5.5. For infinite lists (infinite scroll) use Relay Cursor Connections Specification.</a></li>
</ul>
</li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/06-mutations/README.md">6. Mutation rules</a>
<ul>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/06-mutations/mutation-namespaces.md">6.1. Use Namespace-types to group mutations within a single resource.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/06-mutations/mutation-business-operations.md">6.2. Go beyond CRUD – create small mutations for different business operations on resources.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/06-mutations/mutation-batch-changes.md">6.3. Consider the ability to perform mutations on multiple items (same type batch changes).</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/06-mutations/mutation-required-args.md">6.4. Mutations should clearly describe all the mandatory arguments, there should be no options either-either.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/06-mutations/mutation-input-arg.md">6.5. In mutations, put all variables into one unique input argument.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/06-mutations/mutation-payload.md">6.6. Every mutation should have a unique payload type.</a>
<ul>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/06-mutations/mutation-payload-record.md">6.6.1. In the mutation response, return the modified resource and its <code>id</code>.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/06-mutations/mutation-payload-status.md">6.6.2. Return operation status in mutation response.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/06-mutations/mutation-payload-query.md">6.6.3. In the mutation response, return a field of type <code>Query</code>.</a></li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/06-mutations/mutation-payload-errors.md">6.6.4. In the mutation response, return the <code>errors</code> field with typed user errors.</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/07-relations/README.md">7. Rules of linkages between types (relationships)</a>
<ul>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/07-relations/relations-hairy-graphql.md">7.1. GraphQL schema should be "hairy"</a></li>
</ul>
</li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/08-authorization/README.md">8. Authorization</a>
<ul>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/08-authorization/authorization-schema-diffing.md">8.1. Schema diffing based authorization</a></li>
</ul>
</li>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/10-misc/README.md">10. Other rules</a>
<ul>
<li><a href="/graphql-rules/graphql-rules/blob/master/docs/rules/10-misc/misc-docs-markdown.md">10.1. Use markdown for documentation</a></li>
</ul>
</li>
</ul>

**Comment by nodkz** [comment]
Anyway tags addition is a great idea 👍

In the future, we can add more tags with different meanings and allow users to filter rules by them. Eg. `mutation`, `query`, `filtering`, `auth`, `args`, `convention` etc.

But first, need to choose good values for pattern tagging.

**Comment by SachaG** [comment]
> I agree with basic and advanced but don't feel that experimental name is good enough.

What's the word that other specs use when something is under review or not yet officially accepted? Under consideration? Under review?
